.circleci/packages/config.yaml

```diff
@@ -0,0 +1,54 @@
+
+version:
+  release:
+    match: '^v[0-9]+.[0-9]+.[0-9]+'
+    value: '{{env.CIRCLE_TAG[1:]}}'
+  alpha:
+    match: '^v3.0.0.alpha.[0-9]+.[0-9]+'
+    value: '{{env.CIRCLE_TAG[1:]}}'
+  default:
+    value: '3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}'
```

Comment:
You're using `v3...` above and `3...` here. Update to `v3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}` here?

That said, I'd just assume remove `v` from the other two instead as it adds a complication when building debs and rpms, but perhaps it is make something else easier that I'm unaware of. Cc @bnpfeife

Reply:
When `release.match` matches against `$CIRCLE_TAG`, `release.value` is supplied as the package version. Otherwise, `default.version` is supplied as the package version. In retrospect, the tool could probably consider an array of possibilities instead of just `release` and `default`. However, I think that _just_ `release` and `default` will suffice for our current needs.

The intent of `release.match` is to prevent building packages with versions incompatible with apt/deb and yum/rpm. Unfortunately, our release tags have included the `v` prefix (even since before I started). So, without wanting to cause trouble, I just made the tooling work around it. In most cases, `match` should include the `^v`, but `value` should not. `{{ env.CIRCLE_TAG[1:] }}` instructs the templating engine to strip the first character from `$CIRCLE_TAG`.

With all that being said, I think this is the correct specification:
```yaml
version:
  release:
    match: '^v[0-9]+.[0-9]+.[0-9]+'
    value: '{{env.CIRCLE_TAG[1:]}}'
  default:
    value: '3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}'
```

Both release and alpha tags should retain everything except the `v`. In all other circumstances, `3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}` will be used instead.

Reply:
So this will also work if say we do a tag of `v3.0.0-0.alpha.0.2`?

Reply:
> However, like @mgattozzi said, we want to support versions like 3.0.0-0.alpha.0.1, 3.0.0-0.alpha.1, 3.0.0-0.beta.1 and 3.0.0-0.rc.1 in addition to 3.0.0 (ie, release) and 3.0.0+snapshot-deadbeef (ie, default), but it's not clear (to me!) how those would fit in to what is there now (just default andrelease). 

I think this may 'just work', since `match: '^v[0-9]+.[0-9]+.[0-9]+'` should match all of those and so just end up with the tag, so long as the tag prefixes with `v`. Eg: `3.0.0-0.alpha.0.1`, `3.0.0-0.alpha.1`, `v3.0.0-0.beta.1`, `3.0.0-0.rc.1`, `v3.0.0`.

Reply:
We could be fancier and do:
```
match: '^v[0-9]+.[0-9]+.[0-9]+(-[0-9]+\.(alpha|beta|rc)\.[0-9]+(\.[0-9]+)?)?$'
```

Eg:
```
$ for i in v3.0.0-0.alpha.0.1 v3.0.0-0.alpha.1 v3.0.0-0.beta.1 v3.0.0-0.rc.1 v3.0.0 good-before-bad-after v3.0.0.alpha.0.1 v3.0.0.alpha.1 3.0.0-0.beta.1 v3.0.0-0.beta v3.0.0rc ; do echo "$i" | grep -E '^v[0-9]+.[0-9]+.[0-9]+(-[0-9]+\.(alpha|beta|rc)\.[0-9]+(\.[0-9]+)?)?$' ; done
v3.0.0-0.alpha.0.1
v3.0.0-0.alpha.1
v3.0.0-0.beta.1
v3.0.0-0.rc.1
v3.0.0
```
This would force us to conform to the version if we want a tag to end up in `releases/` that is an alpha, beta, rc or release.

Reply:
Okay, I manually supplied `CIRCLE_*` values to `ci-packager`. It looks like everything works as expected.

```sh
$ docker run -it --rm \
  --env CIRCLE_TAG=v3.0.0-0.alpha.1 \
  --env CIRCLE_SHA1="$(git rev-parse HEAD)" \
  --env CIRCLE_BRANCH="$(git rev-parse --abbrev-ref HEAD)" \
  -v "/tmp/workspace:/tmp/workspace" \
  -v "${PWD}:/project" \
  us-east1-docker.pkg.dev/influxdata-team-edge/ci-support/ci-packager:latest bash
  
$ cd /project && packager .circleci/packager/config.yaml
...

$ ls -l artifacts
total 232400
-rw-r--r-- 1 root root 46982921 Apr 11 20:32 influxdb3-3.0.0-0.alpha.1-windows.zip
-rw-r--r-- 1 root root 24335388 Apr 11 20:32 influxdb3-3.0.0-0.alpha.1_darwin_amd64.tar.gz
-rw-r--r-- 1 root root 28810063 Apr 11 20:32 influxdb3-3.0.0-0.alpha.1_linux_amd64.tar.gz
-rw-r--r-- 1 root root 26864086 Apr 11 20:32 influxdb3-3.0.0-0.alpha.1_linux_arm64.tar.gz
-rw-r--r-- 1 root root 26746749 Apr 11 20:32 influxdb3-3.0.0_0.alpha.1-1.aarch64.rpm
-rw-r--r-- 1 root root 28542862 Apr 11 20:32 influxdb3-3.0.0_0.alpha.1-1.x86_64.rpm
-rw-r--r-- 1 root root 28812320 Apr 11 20:32 influxdb3_3.0.0-0.alpha.1-1_amd64.deb
-rw-r--r-- 1 root root 26863794 Apr 11 20:32 influxdb3_3.0.0-0.alpha.1-1_arm64.de
```

I then verified the package versions outside the docker container:
```
$ dpkg -I influxdb3_3.0.0-0.alpha.1-1_amd64.deb 
 new Debian package, version 2.0.
 size 28812320 bytes: control archive=793 bytes.
     279 bytes,    11 lines      control              
     322 bytes,     5 lines      md5sums              
      12 bytes,     1 lines   *  postinst             #!/bin/bash
     434 bytes,    22 lines   *  preinst              #!/bin/bash
 Package: influxdb3
 Version: 3.0.0-0.alpha.1-1
 License: MIT/Apache-2.0
 Vendor: InfluxData
 Architecture: amd64
 Maintainer: support@influxdb.com
 Installed-Size: 85717
 Section: default
 Priority: optional
 Homepage: https://influxdata.com
 Description: Monolithic time-series database.
```

```
$ rpm -qip influxdb3-3.0.0_0.alpha.1-1.x86_64.rpm
Name        : influxdb3
Version     : 3.0.0_0.alpha.1
Release     : 1
Architecture: x86_64
Install Date: (not installed)
Group       : default
Size        : 87774723
License     : MIT/Apache-2.0
Signature   : (none)
Source RPM  : influxdb3-3.0.0_0.alpha.1-1.src.rpm
Build Date  : Thu 11 Apr 2024 04:32:07 PM EDT
Build Host  : 18795ca60e10
Relocations : / 
Packager    : support@influxdb.com
Vendor      : InfluxData
URL         : https://influxdata.com
Summary     : Monolithic time-series database.
Description :
Monolithic time-series database.
```
I'll test the remaining variants. I'll only add another comment if I encounter breakage. :slightly_smiling_face:

Reply:
> $ dpkg -I influxdb3_3.0.0-0.alpha.1-1_amd64.deb 
> ...
> Package: influxdb3
> Version: 3.0.0-0.alpha.1-1

@bnpfeife - note that with the alpha, beta and rc builds that have a tag of form `v3.0.0-0.alpha.1`, etc, we don't need to tag `-1` on the end, that is what the `-0` part of `-0.alpha.1` is for. We only need to tack on `-1` if the tag is of form `^v[0-9]+.[0-9]+.[0-9]+$` (or doesn't contain 'alpha', 'beta' or 'rc', either works). It's ok that it is there from a package manager point of view, but it is extraneous.

Reply:
@mgattozzi @jdstrand 

Sorry about the confusion between `release` and `default`. I think that `release` is a misnomer and implies that there is something special about the package. The _only_ difference is which `value` is used for the package version. In retrospect, if I were to re-write this portion of `ci-packager`, the `version` section would look like this instead:
```yaml
version:
# Matches "release" versions (e.g v3.0.0)
- match: '^v[0-9]+.[0-9]+.[0-9]+$'
  value: '{{env.CIRCLE_TAG[1:]}}'
# Matches "alpha" versions (e.g v3.0.0-0.alpha)
- match: '^v[0-9]+.[0-9]+.[0-9]+-[0-9].alpha'
  value: '{{env.CIRCLE_TAG[1:]}}'
# Catch-all versions
- match: '.*'
  value: '3.0.0+snapshot-{{env.CIRCLE_SHA1[:8]}}'
```

Reply:
> $ rpm -qip influxdb3-3.0.0_0.alpha.1-1.x86_64.rpm
> Name        : influxdb3
> Version     : 3.0.0_0.alpha.1

@bnpfeife - I expected the rpm build to use a `-` instead of an `_` here in both the file name and the internal version.